### ISSUE TITLE
[scroll-animations-1][web-animations-2] getCurrentTime is self-inconsistent wrt representing time

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#8765]
In order to represent time in relation to named timeline ranges, we have two options:

- Represent as absolute progress from the start of the range, in units consistent with the overall timeline (i.e. equivalent to currentTime).
- Represent as percentage progress between the start and end of the range.

If we consider the default behavior as returning the time with respect to a timeline range that encompasses the entire timeline, `getCurrentTime()` switches between the two modes depending on whether you specify a range name or not. (!)

This doesn't feel unnatural to scroll timelines, because their “absolute timeline units” are percentages of the timeline (which itself has a finite range)... We would probably have noticed the discrepency sooner if they used length-based time. :/

But it runs into problems if we extend the named timeline concept to other types of timelines, such as time-based timelines. These bring in (potentially) two differences:

- Their absolute time is not a percentage, its an absolute unit.
- They might possibly not be finite?

I think we might want to rethink this API. Some possible options:
* Restrict (and define) it as returning percentage progress through a range, rather than also being able to return the global `currentTime`, and name it accordingly as @birtles suggests in #8201.
* Define two APIs, one that returns percentage progress and one that returns absolute progress (i.e. units compatible with `currentTime`) in relation to a named range.
* Define a single API that can return either percentage progress or absolute progress depending on its arguments.

I'm not sure which direction we want to go... @bramus, thoughts?

**Comment by bramus** [comment]
I can see that as an author you’d sometimes want to know the relative “time” _(i.e. a percentage)_ and sometimes you’d want the absolute “time”. We could go with two different methods `getRelative…` and `getAbsolute…` for this, or add an extra argument to the property bag of the existing one. Undecided that the default argument value – `relative` or `absolute` – should be in the latter case.

Side note: seeing that I put time between quotes in the previous paragraph and @fantasai used the word “progress” – and I also hinted at it [here](https://github.com/w3c/csswg-drafts/issues/8201#issuecomment-1493397265) – maybe it’s better to rename the method to `getProgress()`?

**Comment by fantasai** [comment]
@bramus ... so maybe we want parallel `getCurrentTime` and `getCurrentProgress` functions where the former is in .currentTime units and the latter is in percentages (always) and returns null on non-finite ranges?

* Option A: `getCurrentTime()` to return the absolute time offset and `getCurrentProgress()` to get the percentage time offset. Passing `{ range: "name" }` gets the offset / percentage relative to that range.
* Option B: `getCurrentTimeForRange("name")` to return the absolute time offset from the start of the named range, `getCurrentProgressForRange("name")` to return the percentage progress through the named range.

**Comment by bramus** [comment]
/ping @flackr to get your take on this. I think this needs to be tackled before shipping.

**Comment by birtles** [comment]
I agree this should be resolved before shipping. I wonder how important the API even is? Do authors need it? Is it needed for WPT? If we're not sure what the ergonomics/naming/use cases are for this, would it make sense to ship without it?

**Comment by flackr** [comment]
I agree with @birtles that I think this is purely additive to scroll driven animations and doesn't need to be part of the initial API. I'm in favor of shipping without it and adding it later.

**Comment by fantasai** [comment]
So we resolved to add .progress to the Animation API in https://github.com/w3c/csswg-drafts/issues/8799 ; but that still leaves this issue open.

**Comment by birtles** [comment]
So long as we don't have a strong use case (it seems like `.progress` covers the use cases we're aware of) and we don't have a clear picture of what the shape of the API should me, I'm very much in favor of dropping it from Level 1.

**Comment by graouts** [comment]
Renaming the relevant set as `.tentative` in https://github.com/web-platform-tests/wpt/pull/51133.
